reactor_backend: Fix iocb double-free during AIO teardown

tchaikov opened the pull request:
Fix the double-free bug in iocb pool management that causes bad_alloc during AIO teardown (issue #3169, https://github.com/scylladb/scylladb/issues/17369, https://github.com/scylladb/scylladb/issues/26481).

Root Cause:
The bug occurs when stop() is called while AIO retry operations are in progress. The race condition:

1. schedule_retry() A is processing iocbs in _aio_retries
2. stop() calls std::exchange(_pending_aio_retry_fut, ready_future)
3. New I/Os complete with -EAGAIN, added to _pending_aio_retry
4. submit_work() sees need_to_retry()=TRUE && !retry_in_progress() (because we exchanged the future) and starts a NEW schedule_retry() B
5. schedule_retry() A completes, .finally() block runs
6. .finally() calls reap_completions(false) which returns -EAGAIN iocbs to the pool (since allow_retry=false)
7. But schedule_retry() B is still running with those same iocbs in _aio_retries!
8. When B tries to process them and encounters errors, handle_aio_error() calls put_one() again
9. Double-free: attempting to push to a full 1024-capacity free list, causing boost::container::bad_alloc

The Fix:
Add a _stopping flag that is set at the start of stop(). Check this flag in need_to_retry() to prevent new schedule_retry() loops from starting during shutdown. This eliminates the race window where a second retry loop could start after the future is exchanged.

The existing retry loop completes normally, and stop()'s .finally() block simply waits for all in-flight iocbs to be reaped and returned. No iocbs are leaked, and no double-free can occur.

Additionally, add debug-mode tracking (_iocb_allocated[]) to detect future double-free bugs and provide diagnostic information.

Fixes: #3169

tchaikov committed:
reactor_backend: fix iocb double-free and shutdown hang during AIO teardown

We had a race where stop() could be called while schedule_retry() was in
progress, letting a second retry loop start and both loops processing the
same iocbs -> double-free, crash.

The root cause: when stop() exchanges _pending_aio_retry_fut for a ready
future, retry_in_progress() becomes false. If submit_work() fires before
the first loop finishes, it starts a second schedule_retry(). Both loops
then race on _aio_retries and _pending_aio_retry, calling put_one() on
the same iocbs twice.

The fix is to gate schedule_retry() on a new _stopping flag so only one
loop can ever run. stop() sets _stopping first, waits for any in-flight
loop via .finally(), then drains all outstanding iocbs in a do_until
loop that calls reap_completions() and reap_pending_retries() each
iteration. reap_completions() forces allow_retry=false when _stopping so
EAGAIN iocbs complete immediately instead of accumulating. submit_work()
stays unchanged as stop() owns cleanup of whatever arrives during
shutdown.

Also adds SEASTAR_IOCB_POOL_DEBUG (a std::bitset<max_aio>) to catch
double-alloc or double-free of iocbs in debug builds.

Fixes #3169
Signed-off-by: Kefu Chai <k.chai@proxmox.com>

xemul commented on src/core/reactor_backend.cc:
```diff
@@ -115,6 +115,9 @@ aio_storage_context::iocb_pool::iocb_pool() {
     for (unsigned i = 0; i != max_aio; ++i) {
         _free_iocbs.push(&_all_iocbs[i]);
     }
+#ifndef NDEBUG
```
Please, define SEASTAR_IOCB_POOL_DEBUG, like e.g. SEASTAR_GATE_HOLDER_DEBUG

tchaikov replied:
sure. defined.

tchaikov commented on src/core/reactor_backend.hh:
```diff
@@ -64,6 +64,11 @@ class aio_storage_context {
+#ifndef NDEBUG
+        // Track which iocbs are currently allocated (not in free list)
+        // This helps detect double-free bugs
+        std::array<bool, max_aio> _iocb_allocated;
```
indeed. switched to `std::bitset`.

tchaikov commented:
> > * schedule_retry() A is processing iocbs in _aio_retries
> 
> OK, so in the picture, there's iocb_X that gets into the kernel.
> 
> > * stop() calls std::exchange(_pending_aio_retry_fut, ready_future)
> > * New I/Os complete with -EAGAIN, added to _pending_aio_retry
> 
> In the picture, there's iocb_Y that completes with EAGAIN and gets into _pending_aio_retry
> 
> > * submit_work() sees need_to_retry()=TRUE && !retry_in_progress() (because we exchanged the future) and starts a NEW schedule_retry() B
> 
> In the picture, it says that iocb_Y gets into _aio_retries
> 
> > * schedule_retry() A completes, .finally() block runs
> > * .finally() calls reap_completions(false) which returns -EAGAIN iocbs to the pool (since allow_retry=false)
> 
> In the picture, this place says that iocb_Y completes with EAGAIN. If iocb_Y completed, it means that it was submitted in previous step by schedule_retry() B, is that correct?
> 
> > * But schedule_retry() B is still running with those same iocbs in _aio_retries!
> > * When B tries to process them and encounters errors, handle_aio_error() calls put_one() again
> 
> But why? The schedule_retry() B calls io_submit(iocb_Y) and gets error (picture says it's EBADF). In that case, how could reap_completions(false) see that iocb, if it wasn't submitted by schedule_retry() (it returned EBADF on submission). At the same time, if iocb_Y got to that place, it means that it was submitted by schedule_retry() B without errors.


  @xemul You're absolutely right to question this  - the specific sequence I described may not be entirely accurate. Let me clarify just to make sure we are now on the same page:

  ## The Actual Race Condition

  The core issue is simpler than my original description: **two concurrent `schedule_retry()` loops can run simultaneously during shutdown**, which leads to shared state corruption in `_aio_retries` and `_pending_aio_retry` vectors.

  Here's how it happens:

  1. **`schedule_retry() A`** is running: it has swapped `_aio_retries` with `_pending_aio_retry` and is in the middle of submitting iocbs via the thread pool
  2. **`stop()` is called**: executes `std::exchange(_pending_aio_retry_fut, make_ready_future<>())`
  3. **`submit_work()` gets called** by the poller (which is still running)
  4. **Line 277 check**: `retry_in_progress()` now returns **false** (because we just exchanged the future), but there are still items in the retry queues
  5. **`schedule_retry() B` starts**: now we have TWO loops both manipulating the same `_aio_retries` and `_pending_aio_retry` vectors

  ## The Double-Free

  The double-free doesn't necessarily happen through the completion path - it happens because:

  - Both loops perform `_aio_retries.erase(...)` operations (line 323)
  - Both loops perform `std::swap(_aio_retries, _pending_aio_retry)` (line 294)
  - These operations on shared state can cause the same iocb to be erased/processed twice
  - When iocbs are "consumed" and erased, they get returned to the pool via `handle_aio_error()` or via the completion handler
  - If both loops think they consumed the same iocb, it gets returned to the pool twice = double-free

> 
> > * Double-free: attempting to push to a full 1024-capacity free list, causing boost::container::bad_alloc

tchaikov commented:
v3:

- do not check for `_stopping` in `need_to_retry()`
- incorporated Pavel's suggestion in `aio_storage_context::submit_work()`

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -131,7 +134,17 @@ aio_storage_context::~aio_storage_context() {
+        desc->complete_with(-ECANCELED);
+        return true;
+    });
+
```
Q: why are we even here? If we're shutting down the reactor, nothing should be in progress.

tchaikov reviewed (commented):
v4:

- dropped the defensive draining logic in `aio_storage_context::stop()`
- guarded the assert of `_free_iocbs.size() < max_aio` in `aio_storage_context::iocb_pool::put_one()` with `SEASTAR_IOCB_POOL_DEBUG`, as it's in the hot path.

xemul commented on src/core/reactor_backend.cc:
```diff
@@ -199,6 +238,10 @@ aio_storage_context::handle_aio_error(linux_abi::iocb* iocb, int ec) {
 
 bool
 aio_storage_context::submit_work() {
+    if (_stopping) [[unlikely]] {
```
Let's allow `submit_work()` do whatever it prefers and make `aio_storage_context::stop()` collect whatever appears from it. At most, add extra activity under existing `need_retry()` check.

Maybe something like that would work? (I'm not sure that putting default-initialized future into `_pending_aio_retry_fut` is good)

```patch
diff --git a/src/core/reactor_backend.cc b/src/core/reactor_backend.cc
index 725f0daf6..4e6dc20bf 100644
--- a/src/core/reactor_backend.cc
+++ b/src/core/reactor_backend.cc
@@ -131,9 +131,10 @@ aio_storage_context::~aio_storage_context() {
 }
 
 future<> aio_storage_context::stop() noexcept {
-    return std::exchange(_pending_aio_retry_fut, make_ready_future<>()).finally([this] {
+    return std::exchange(_pending_aio_retry_fut, future<>() /* not future::available() future */).finally([this] {
         return do_until([this] { return !_iocb_pool.outstanding(); }, [this] {
             reap_completions(false);
+            reap_pending_retries();
             return make_ready_future<>();
         });
     });
@@ -243,7 +244,7 @@ aio_storage_context::submit_work() {
         did_work = true;
     }
 
-    if (need_to_retry() && !retry_in_progress()) {
+    if (need_to_retry()) {
         schedule_retry();
     }
 
@@ -251,6 +252,10 @@ aio_storage_context::submit_work() {
 }
 
 void aio_storage_context::schedule_retry() {
+    if (!_pending_aio_retry_fut.available()) {
+        return;
+    }
+
     // loop until both _pending_aio_retry and _aio_retries are empty.
     // While retrying _aio_retries, new retries may be queued onto _pending_aio_retry.
     _pending_aio_retry_fut = do_until([this] {
@@ -317,6 +322,16 @@ bool aio_storage_context::reap_completions(bool allow_retry)
     return n;
 }
 
+void aio_storage_context::reap_pending_retries()
+{
+    for (auto* iocb : _pending_aio_retry) {
+        auto desc = get_user_data<kernel_completion>(*iocb);
+        desc->complete_with(-ECANCELED);
+        _iocb_pool.put_one(iocb);
+    }
+    _pending_aio_retry.clear();
+}
+
 bool aio_storage_context::can_sleep() const {
     // Because aio depends on polling, it cannot generate events to wake us up, Therefore, sleep
     // is only possible if there are no in-flight aios. If there are, we need to keep polling.
diff --git a/src/core/reactor_backend.hh b/src/core/reactor_backend.hh
index aa240b8c6..3d4efb9c0 100644
--- a/src/core/reactor_backend.hh
+++ b/src/core/reactor_backend.hh
@@ -87,9 +87,7 @@ class aio_storage_context {
         return !_pending_aio_retry.empty() || !_aio_retries.empty();
     }
 
-    bool retry_in_progress() const noexcept {
-        return !_pending_aio_retry_fut.available();
-    }
+    void reap_pending_retries();
 
 public:
     explicit aio_storage_context(reactor& r);

```

avikivity commented:
> The existing retry loop completes normally, and stop()'s .finally() block simply waits for all in-flight iocbs to be reaped and returned. No iocbs are leaked, and no double-free can occur.

This means the application leaked futures, no?

In general, if the application leaks work-in-progress, we can't recover.

xemul commented:
Maybe the race is that both, app_template::run() and aio_storage_context::aio_storage_context() both schedule at_exit callbacks? The run's exit happens after aio context's one, so by the time user func is trying to collect all its futures, the aio context's cleanup had been executed already and _pending_aio_retry_fut had been exchanged with dummy ready one thus opening the window for two concurrent retry fibers

emdroid commented:
Hi, what is the status of this fix? This seems to be fixing a recurring CI-stability issue:
* [SCYLLADB-302](https://scylladb.atlassian.net/browse/SCYLLADB-302)
* [SCYLLADB-1530](https://scylladb.atlassian.net/browse/SCYLLADB-1530) , [SCYLLADB-1669](https://scylladb.atlassian.net/browse/SCYLLADB-1669), [SCYLLADB-1866](https://scylladb.atlassian.net/browse/SCYLLADB-1866) (new duplicates)

[SCYLLADB-302]: https://scylladb.atlassian.net/browse/SCYLLADB-302?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SCYLLADB-1530]: https://scylladb.atlassian.net/browse/SCYLLADB-1530?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SCYLLADB-1669]: https://scylladb.atlassian.net/browse/SCYLLADB-1669?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SCYLLADB-1866]: https://scylladb.atlassian.net/browse/SCYLLADB-1866?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

tchaikov commented:
@emdroid  Thanks for the ping. i just reworked the change and rewrote the commit message in hope to address the questions from the reviewers.

v5:

- while re-tracing the code paths, i also found that the earlier version left a follow-on shutdown hang under `--kernel-page-cache` (every iocb is routed through `_pending_aio_retry` there, and once `schedule_retry()` is gated during shutdown nothing drains it), so this revision closes that too.

xemul commented:
@tchaikov , [there's](https://github.com/scylladb/seastar/pull/3182#issuecomment-4262551874) still a concern

> This means the application leaked futures, no?

I partially [second](https://github.com/scylladb/seastar/pull/3182#issuecomment-4278000930) one: maybe the problem is that aio_storage_context::stop() should be called later?

xemul commented on src/core/reactor_backend.cc:
```diff
@@ -130,8 +133,45 @@ aio_storage_context::~aio_storage_context() {
+        f.ignore_ready_future();
+        reap_pending_retries();
         return do_until([this] { return !_iocb_pool.outstanding(); }, [this] {
             reap_completions(false);
```
This `false` seem not to be needed any longer. The `_stopping` is set, so reap_completions() would disable retries on its own

tchaikov replied:
yup, dropped.

xemul commented on src/core/reactor_backend.cc:
```diff
@@ -201,6 +258,22 @@ bool
 aio_storage_context::submit_work() {
     bool did_work = false;
 
+    // During shutdown, cancel any new I/O instead of allocating iocbs.
```
Error handling at the beginning of the happy path is weird. What if we process normally and do submit these IOs while _stopping == true? Isn't there already a code that suggests to pick it up? Or this path is not supposed to happen on stop when _stopping == true?

tchaikov replied:
indeed. dropped.

xemul commented on src/core/reactor_backend.cc:
```diff
@@ -130,8 +133,45 @@ aio_storage_context::~aio_storage_context() {
+    // Use then_wrapped so the drain runs even if the retry loop future holds
+    // an exception.
+    return std::exchange(_pending_aio_retry_fut, make_ready_future<>()).then_wrapped([this] (future<> f) noexcept {
+        f.ignore_ready_future();
```
Why is it necessary to replace .finally() with .then_wrapped + ignore_ready_future ?

tchaikov replied:
ahh, yes. it's a leftover of an experiment. it's not necessary. reverted.

mykaul commented:
@tchaikov - hi - are you still working on this?
@emdroid - if the answer to the above is negative, please own it.

tchaikov commented:
> @tchaikov , [there's](https://github.com/scylladb/seastar/pull/3182#issuecomment-4262551874) still a concern
> 
> > This means the application leaked futures, no?
> 
> I partially [second](https://github.com/scylladb/seastar/pull/3182#issuecomment-4278000930) one: maybe the problem is that aio_storage_context::stop() should be called later?

i am not sure if i am following you, but `aio_storage_context::stop()` is actually already called last, because the AIO context registers it with `_r.do_at_exit()`. since `run_exit_tasks()` iterates that list in reverse, `stop()` fires after all application cleanup.

tchaikov commented:
v6:

- addressed @xemul's comments.

tchaikov commented:
> @tchaikov - hi - are you still working on this? @emdroid - if the answer to the above is negative, please own it.

still on it. just pushed v6 addressing Pavel's comments. happy for anyone to pile on.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -130,10 +133,43 @@ aio_storage_context::~aio_storage_context() {
+    for (auto iocb : _aio_retries) {
+        _iocb_pool.put_one(iocb);
+        auto desc = get_user_data<kernel_completion>(*iocb);
+        desc->complete_with(-ECANCELED);
```
This will probably unleash a bunch of latent bugs, because most code will be unprepared for these errors.

It might be better to let them complete naturally.

We can try with cancellation first, it won't be worse than the crash.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -243,14 +296,23 @@ aio_storage_context::submit_work() {
+    //       queues itself; starting a second loop here would race with
+    //       stop()'s drain on the same iocb pool (the original double-free).
+    if (!_pending_aio_retry_fut.available() || _stopping) {
+        return;
```
What if it's not available now, but is in the process of becoming available?

avikivity replied:
For example, do_until was preempted and is waiting to be scheduled. Might not actually be possible, since it will re-evaluate the condition. Still, it's not nice to depend on such details.

avikivity replied:
But, I see it's a preexisting problem.

More modern is to start a retry fiber in the constructor, wake it each time something needs to be done, and collect it in stop().

bhalevy replied:
Agree with the above.
With a fiber design, the cancellation of the retried iocb's can be done by the retry fiber when it sees `_stopping` rather than special handled in the stop() path.
Maybe it's a good opportunity to do that in the scope of this change.

bhalevy replied:
See https://github.com/scylladb/seastar/pull/3408

avikivity reviewed (approved):
Looks good but will let @xemul make the final call.

bhalevy commented on src/core/reactor_backend.cc:
```diff
@@ -130,10 +133,43 @@ aio_storage_context::~aio_storage_context() {
+        auto desc = get_user_data<kernel_completion>(*iocb);
+        desc->complete_with(-ECANCELED);
+    }
+    _aio_retries.clear();
```
nit: since the logic for reaping `_pending_aio_retry` and `_aio_retries` is identical, can we implement it in a shared function, called for each queue?

tchaikov replied:
created #3406

bhalevy replied:
See https://github.com/scylladb/seastar/pull/3408

bhalevy commented on src/core/reactor_backend.cc:
```diff
@@ -296,6 +358,11 @@ void aio_storage_context::schedule_retry() {
+    // During shutdown, complete EAGAIN iocbs immediately instead of
+    // queuing them for retry — schedule_retry() is gated on _stopping.
+    if (_stopping) {
+        allow_retry = false;
```
nit: `allow_retry &= !_stopping` rather than adding a conditional

tchaikov replied:
it's not in the hot paths, so i'd prefer trading a condition statement for better readability.

bhalevy replied:
I'm working on a cleanup series that addresses this in even a simpler way.
I'll post a link once pushed

bhalevy replied:
See https://github.com/scylladb/seastar/pull/3408 - the allow_retry parameter is removed there altogether and when _stopping we cancel the iocb in the same way it is handled by the retry_fiber so we don't end up with 2 different states of canceled retries on shutdown
